### PR TITLE
Implementation of a custom media_player entity for linked input

### PR DESCRIPTION
**Pull request by dan-s-github**
### Core Functionality
- **Input Proxy Entities**: Creates media_player entities for each configured Triad AMS input that proxy linked source media players
- **Device Class**: `RECEIVER` - properly categorized for Home Assistant UI
- **Entity Naming**: Uses input names from the Triad AMS configuration

### State & Availability Management
- **State Proxying**: Proxies ON/OFF state from linked source media player
  - Returns OFF when no linked entity is configured
  - Treats "unknown" state as OFF
- **Availability Tracking**: Silver quality requirement with `PARALLEL_UPDATES = 1`
- **Coordinator Integration**: Subscribes to availability changes from the coordinator

### Feature Management
- **Smart Feature Masking**: Proxies all supported features from linked entity EXCEPT volume controls
  - Removes `VOLUME_SET`, `VOLUME_MUTE`, `VOLUME_STEP` flags
  - Adds `GROUPING` support for multi-room audio
  - Rationale: Volume should only be controlled on physical outputs/speakers

### Media Information Proxying
Proxies all media metadata from linked source entity:
- `media_title`, `media_artist`, `media_album_name`
- `media_duration`, `media_content_id`, `media_content_type`
- `entity_picture` (artwork)
- `shuffle`, `repeat` state
- `sound_mode`, `sound_mode_list`

### Playback Control Commands
All commands proxy to linked entity with non-blocking service calls:
- `async_media_play`, `async_media_pause`, `async_media_stop`
- `async_media_next_track`, `async_media_previous_track`
- `async_media_seek` (with position parameter)
- `async_play_media` (with media type and ID)
- `async_select_source` (source selection)
- `async_shuffle_set` (with keyword-only shuffle parameter)
- `async_set_repeat` (off/all/one)

### Grouping Features
- **Domain-Based Grouping**: `async_join_players` implements intelligent grouping
  - Validates all members are from the integration
  - Supports cross-domain grouping (delegates to linked entity if it supports grouping)
  - Routes Triad AMS outputs to the input source
  - Updates group_members list on all affected entities
- **Custom Exception**: `InvalidGroupMemberError` with translation support for invalid group operations
- **Group Member Merging**: Combines local output members with linked entity's group members

### Entity Lifecycle
- **State Change Tracking**: Subscribes to linked entity state changes
- **Proper Cleanup**: Unsubscribes from state changes and availability listeners on removal
- **Stale Entity Cleanup**: Automatically removes entities for disabled inputs

### Testing Coverage
- **17 unit tests** in test_input_media_player.py:
  - Feature masking validation
  - State proxying (off/linked/unknown)
  - Group member merging and missing state handling
  - 11 playback command proxy tests
- **11 Silver quality tests** in `test_input_media_player_silver.py`:
  - Availability tracking (5 tests)
  - Logging requirements (4 tests)
  - PARALLEL_UPDATES constant (2 tests)

All 282 tests passing ✅

### TODO
- Implement service method `async_get_join_candidates` that would allow media-player cards to request a curated list of joinable media_players. 
-- feature request in HA Frontend to include this option and eventually a patch so that it would work with the default 'media control' card
- Implement Gold quality requirements
- Review AI generated Silver quality test cases

Screenshots
image
image

### CODE REVIEW

**Comment by bharat** [comment]
I'm psyched for this! This PR is still marked as draft so I'll hold off on reviewing it in depth, but here's what my AI had to say about it:

# Code Review: `dev/input-media-player-entities` Branch

## Summary
This branch adds input media player entities that proxy linked `media_player` entities, with full grouping support. The implementation adds **+1452/-46 lines** across 13 files, introducing a new `TriadAmsInputMediaPlayer` entity class and comprehensive test coverage.

## ✅ Strengths

### 1. **Architecture & Design**
- Clean separation of concerns: `TriadAmsInput` model class, `TriadAmsInputMediaPlayer` entity class
- Well-structured helper functions (`_build_inputs`, `_create_input_entities`, `_populate_group_members`) improve code readability
- Proper proxying of media attributes and playback commands to linked entities

### 2. **Test Coverage**
- New comprehensive test files: `test_input_media_player.py`, `test_input_media_player_silver.py`
- Tests cover feature masking, state handling, group membership, command proxying, and Silver requirements
- Existing tests updated appropriately to handle new input entities

### 3. **Home Assistant Integration**
- Proper use of `MediaPlayerEntityFeature.GROUPING`
- Correct implementation of `group_members` and `group_leader` properties
- Availability tracking properly implemented for Silver requirements
- `PARALLEL_UPDATES = 1` correctly set

### 4. **Error Handling**
- New `InvalidGroupMemberError` exception with proper translation support
- Good validation in `async_join_players` method
- Translation strings added to `en.json`

## ⚠️ Issues and Concerns

### 🔴 Critical Issues

#### 1. **Missing Coordinator Reference in `TriadAmsInput`**
```python
# In TriadAmsInputMediaPlayer.available property:
coordinator = getattr(self.input, "coordinator", None)
```
The `TriadAmsInput` model doesn't store a coordinator reference, but the code expects it. This will fail at runtime when checking availability.

**Fix needed:**
- Add `coordinator` parameter to `TriadAmsInput.__init__`
- Or pass coordinator when needed through a different mechanism
- Update `_build_inputs` to pass coordinator

#### 2. **Group Members List Mutation Logic**
```python
# In async_join_players:
self._group_members = [
    m for m in group_members if m not in self._group_members
] + self._group_members
```
This prepends new members but doesn't remove members that are no longer in `group_members`. The join operation should replace the list, not merge.

**Suggested fix:**
```python
# Replace the entire list, not merge
self._group_members = list(group_members)
```

#### 3. **Type Inconsistency: `input_links` Parameter**
The parameter type changed from `dict[int, str | None]` to `dict[str, str | None]` in `TriadAmsMediaPlayer.__init__`. While the comment explains this (options store keys as strings), ensure all call sites are updated consistently.

### 🟡 Medium Priority Issues

#### 4. **Incomplete Method Implementation**
The diff shows `async_play_media` ending with `blocking=Fa` (truncated). Verify it's `blocking=False` in the actual code.

#### 5. **Test Coverage Gaps**
- Missing tests for `async_join_players` with mixed domains
- Missing error case tests when linked entity doesn't support grouping
- Edge cases when `group_members` is None vs empty list

### 🟢 Minor Issues

#### 6. **Code Duplication**
Similar patterns in `_linked_attr` methods could be consolidated. Consider a base class or mixin for shared proxy behavior.

#### 7. **Documentation**
- Add docstrings explaining the grouping model
- Document the relationship between input entities and output entities
- Explain when input entities are created vs when they're skipped

#### 8. **Performance Considerations**
- `async_join_players` uses `asyncio.gather` correctly
- Consider batching registry lookups if many members are processed

#### 9. **Logging**
Good use of logging in availability transitions. Consider adding debug logs for group membership changes.

## 📋 Specific Code Review Points

### 1. `_find_linked_input` Function
```python
def _find_linked_input(
    _output: TriadAmsOutput,
    input_links_opt: dict[str, str],
    input_entities: dict[int, TriadAmsInputMediaPlayer],
) -> str | None:
```
The `_output` parameter is unused. Consider removing it or using it for validation.

### 2. `async_turn_on_with_source` Enhancement
Good addition to handle input entity IDs directly. The regex pattern `r"_input_(\d+)$"` is appropriate. Consider caching registry lookups if called frequently.

### 3. Cleanup Function Update
`_cleanup_stale_entities` correctly handles both input and output entities. Tests updated appropriately.

### 4. Repairs Module Refactoring
Change from platform to direct function call is cleaner. Tests updated correctly.

## 🔧 Recommendations

### High Priority
1. ✅ Fix coordinator reference in `TriadAmsInput`
2. ✅ Fix group members list update logic in `async_join_players`
3. ✅ Verify all async methods are complete (no truncation)

### Medium Priority
1. Add integration tests for grouping scenarios
2. Add error handling tests for edge cases
3. Consider adding type hints for all new functions

### Low Priority
1. Refactor common proxy patterns to reduce duplication
2. Add more detailed docstrings
3. Consider performance optimizations for registry lookups

## 📊 Overall Assessment

**Rating: 8/10** (would be 9/10 after fixing the critical issues)

This branch provides a solid foundation for input media player entities with grouping support. The architecture is sound, test coverage is good, and Home Assistant integration follows best practices. The main concerns are around the coordinator reference and group membership list handling, which should be addressed before merging.

The implementation correctly:
- ✅ Proxies media attributes from linked entities
- ✅ Handles grouping with output entities
- ✅ Implements Silver quality scale requirements
- ✅ Provides comprehensive test coverage
- ✅ Follows Home Assistant patterns and conventions

## Questions for Discussion

1. Should input entities be created for all active inputs, or only those with linked entities? (Current implementation only creates for linked inputs - seems correct)
2. How should group membership be updated when an output is manually disconnected? (Currently handled via `async_unjoin_player`)
3. Should there be a limit on the number of group members? (Not currently enforced)

**Comment by bharat** [comment]
Out of the box it seems to work for me. Here's what my Overview looks like with both the Sonos Main and the Triad AMS Main. Note that the Triad AMS shows a total of 6 in the grouping (which is the same as the Sonos).
image

Here's the Sonos group dialog:
image

Here's the Triad group dialog:
image

When I group in Triad AMS Output 3 it seems to do the right thing:
```
2026-01-03 17:54:59.707 INFO (MainThread) [custom_components.triad_ams.connection] Set output 3 to input 3 (resp: Set Out[3] Input Source to input 3)
2026-01-03 17:54:59.711 DEBUG (MainThread) [custom_components.triad_ams.media_player] Main - Current group members: ['media_player.main', 'media_player.downstairs_laundry', 'media_player.lounge', 'media_player.dining_room_bathroom', 'media_player.downstairs_bathroom', 'media_player.secret_bathroom', 'media_player.triad_ams_output_3']
2026-01-03 17:54:59.711 DEBUG (MainThread) [custom_components.triad_ams.media_player] Main - Adding linked group members: ['media_player.main', 'media_player.downstairs_laundry', 'media_player.lounge', 'media_player.dining_room_bathroom', 'media_player.downstairs_bathroom', 'media_player.secret_bathroom']
```

What's interesting is that the number of grouped players now goes to 13:
image

I haven't dug into why that might be happening.

**Comment by dan-s-github** [comment]
updated readme and are no wondering if an input entity should be created even if not linked as it is actually possible to connect anything to an input which can not be controlled by HA. which will then just ignore the proxy stuff, but that could be a future improvement ticket

File: custom_components/triad_ams/input_media_player.py [hunk 1]
```diff
@@ -0,0 +1,775 @@
...
+            entry = registry.async_get(self.input.linked_entity_id)
+            if not entry:
+                return None
+
+            # Get the entity object from the platform (entity_id -> entity object)
+            # This requires accessing the entity registry's internal entities
+            entity_id = self.input.linked_entity_id
+            entity = None
+
+            # Try to find entity in hass.data if it's available
```
Comment (bharat) [inline custom_components/triad_ams/input_media_player.py]: What's the motivation for this approach? It doesn't work for me... it just shows the triad players. 
image

The more idiomatic approach is to use `entity = self.hass.entities.get(entity_id)` which does work for me. I'm assuming that didn't work for you? I can update the code to try that first and then fall back to your approach if you think that makes sense.

This is what I get with my approach:
image

**Comment by dan-s-github** [comment]
Home Assistant Core feature request discussion: https://github.com/orgs/home-assistant/discussions/2342

**Review by bharat — COMMENTED** [review]
I'm back to work so my time is a little limited, but I'm still pushing on this. In the meantime I'll just drop comments one at a time as I encounter issues.

File: custom_components/triad_ams/input_media_player.py (hunk 1 above)
Comment (dan-s-github) [reply custom_components/triad_ams/input_media_player.py]: let my AI review the service implementation it did in the first place, looks bit cleaner now. let me know if this now also works for you

**Comment by bharat** [comment]
That sounds sensible to me. Let me know when you have a branch up and I'll help you test it!

**Comment by bharat** [comment]
Cool. I'll try it out in my dev environment when I get a chance, hopefully later today.

Regarding the naming... I don't see a clear convention. There's `ATTR_GROUP_MEMBERS` which lets you know what's currently in the group, and then there's `SERVICE_JOIN` which really is the action to join the group. I think what's happening here is that we are getting a list of `players` that we can `join` resulting in a `group`. So perhaps the right thing would be `SERVICE_GET_JOINABLE_PLAYERS`?

**Comment by dan-s-github** [comment]
there is now also a frontend branch which uses the new service
https://github.com/dan-s-github/homeassistant-frontend/tree/feature/2342-use-get-groupable-players

regarding naming I will add a comment to the feature request.

**Comment by dan-s-github** [comment]
currently using `install.update` service to update triad_ams to a specific commit

**Comment by dan-s-github** [comment]
as the core pr [changes the entity model](https://developers.home-assistant.io/docs/core/entity/#changing-the-entity-model) a [home assistant architecture issue](https://github.com/home-assistant/architecture/discussions/1331) had to be created to get approval for this change

**Comment by dan-s-github** [comment]
Hi @bharat, I posted in the Discord developers channel without success. 
I also asked in [mediocre-hass-media-player-cards](https://github.com/antontanderup/mediocre-hass-media-player-cards/discussions/389) if this would be useful. And the Codeowner also think this would be nice to have. Maybe he is vote Nr.4
It might be good to create a post about the Triad-AMS integration and the discovered missing functionality in the current media_player implementation when using an Audio Matrix System. 
Hard to tell in which Category it should go.
- Development
- Share your Projects > Custom Integrations

**Comment by gadgetbazza** [comment]
Hi gents, not sure if I'm doing this right, but I've downloaded your pull request and deployed to my HA instance (not a dev setup), and created a new dashboard and added my "Dinine" output zones as a media player to test.

When I then use the speaker group option and add the "Kitchen" zone and press apply, I get an error as per the log below:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 278, in handle_call_service
    response = await hass.services.async_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<7 lines>...
    )
    ^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2817, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2860, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 832, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
        hass, entity, func, data, call.context
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 904, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/media_player/__init__.py", line 1186, in async_join_players
    await self.hass.async_add_executor_job(self.join_players, group_members)
  File "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 59, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/media_player/__init__.py", line 1182, in join_players
    raise NotImplementedError
NotImplementedError

Screenshot 2026-02-21 at 11 38 46

Am I missing something?

Thanks
Barry

**Comment by bharat** [comment]
> "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 59, in run result = self.fn(*self.args, **self.kwargs) File "/usr/src/homeassistant/homeassistant/components/media_player/**init**.py", line 1182, in join_players raise NotImplementedError NotImplementedError

I haven't looked through this in detail, but I'm guessing it's because you're using a stock HomeAssistant and you need one that has our modifications in https://github.com/home-assistant/core/pull/161656 applied. @dan-s-github, does that sound right to you?

**Comment by dan-s-github** [comment]
It's been a while since I was here, unfortunately there is no activity on https://github.com/home-assistant/architecture/discussions/1331 but fixed conflicts anyway

**Comment by bharat** [comment]
Thanks for keeping this fresh and for resolving the conflicts. Still gated on home-assistant/architecture#1331 before this can move, unfortunately, but I appreciate you sticking with it. The AMS24 datapoint on #137 is helpful too.
